Renamed DNSArecord to DNSRecord and included DNS C-Names

DdenBraver opened the pull request:
Updated DSC Resource to include DNS C-names.

msftclas commented:
Hi __@DdenBraver__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@DdenBraver__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

HemantMahawar commented:
@DdenBraver @iainbrighton  @KarolKaczmarek This is a breaking change. On PowerShell gallery, xDnsServer module has been downloaded over **1350** times. Not sure how many of them are using this resource from the module.
Can we gather some data that it is okay to introduce a breaking change?

iainbrighton commented:
I know that a single resource makes sense. However, could we add the renamed resource whilst leaving the DNSArecord intact and add a deprecated note in the readme? Have we done this with any other resources?

This is essentially forking that resource. Whilst it might increase the work in the short-term, it'll probably be worth it in the long run. After all there are more than just A and CNAME records!

> On 12 Nov 2015, at 18:11, Hemant Mahawar <notifications@github.com> wrote:
> 
> @DdenBraver @iainbrighton @KarolKaczmarek This is a breaking change. On PowerShell gallery, xDnsServer module has been downloaded over 1350 times. Not sure how many of them are using this resource from the module.
> Can we gather some data that it is okay to introduce a breaking change?
> 
> —
> Reply to this email directly or view it on GitHub.
>

DdenBraver commented:
@HemantMahawar I fully agree with @iainbrighton. Please let me know if you would like me to add the DNSArecord back to this pull request.
I already have it on my to-do list to update the module for other dns record types.

iainbrighton commented:
@DdenBraver I now need to be able to create multiple A records as well as CNAME records. What is happening with this pull request? I wouldn't want to duplicate any of this work if this can be merged /cc @TravisEz13 @HemantMahawar

iainbrighton commented:
Could we not change the xDnsARecord implementation to a composite resource that simply calls the new xDnsRecord? That way we won't break backwards compatibility and still only have one code base to maintain?

DdenBraver commented:
I am not 100% certain if adding the xDNSARecord as a composite resource inside the module is the best way to go forward, as this requires additional changes to files that are otherwise un-touched.

I can just copy the xDNSRecord code over to xDNSARecord and add a write-warning with a message like: "DSC Resource xDNSARecord has been replaced by xDNSRecord".
Or I can copy back the old xDNSARecord directory and add a warning message.

@TravisEz13 @iainbrighton @HemantMahawar Please let me know what you think is best.

HemantMahawar commented:
@DdenBraver @iainbrighton @TravisEz13 ... Here are my thoughts
I think, having a single resource for CName and ARecord makes sense. We can do the following 2 things to existing DNSARecord resource
- Write a warning, as suggested above, that it is deprecated and replaced by DNSRecord
-  **IF we still want to invest & add to DNSARecord resource**, the Get/Set/Test-TargetResource of that resource can call the Get/Set/Test-TargetResource of DNSRecord resource.

I personally, would leave DNSARecord as is with addition of a warning message and invest only in DNSRecord going forward. That way we can move people over to DNSRecord resource. Also it might be useful to add a warning on module load, so it shows up on calls to Get-DscResource ...

HemantMahawar commented on DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.schema.mof:
```diff
@@ -0,0 +1,9 @@
+{
+    [Key] string Name;
+    [Key] string Zone;
+    [Required,ValueMap{"A-record", "C-name"},Values{"A-record", "C-name"}] string Type;
```
Should it be CName and ARecord instead? That's what I see online and in powershell cmdlets

HemantMahawar commented on DSCResources/MSFT_xDnsRecord/MSFT_xDnsRecord.psm1:
```diff
@@ -46,12 +60,25 @@ function Set-TargetResource
+    if ($Type -eq "A-record")
+    {
+        Write-Verbose "Creating $Type for DNS $Target in $Zone"
+        Add-DnsServerResourceRecordA -IPv4Address $Target -Name $Name -ZoneName $Zone
```
Would it be better to use [Add-DnsServerResourceRecord](https://technet.microsoft.com/en-us/library/jj649925.aspx) cmdlet instead along with splatting so adding more types later will not bloat the code. Something like
```powershell
$parameters = @{IPv4Address = $Target; Name=$Name; ZoneName=$Zone} 
```
and in the if/else decide to add -A or -CName switch parameters as
```powershell 
if($type -eq 'CName'){$parameters.Add('CNAME',$true) } 
```
and outside of if/else splat to [Add-DnsServerResourceRecord](https://technet.microsoft.com/en-us/library/jj649925.aspx)  cmdlet as 
```powershell 
Add-DnsServerResourceRecord @parameters
```

HemantMahawar commented on README.md:
```diff
@@ -30,10 +30,12 @@ Secondary zones allow client machine in primary DNS zones to do DNS resolution o
+### xDnsRecord
 * **Name**: Name of the DNS zone
 * **Zone**: The name of the zone to create the record in
 * **Target**: The IP address of the A record
```
We should clarify that the Target takes **ONLY IPv4 addresses** in current releas

iainbrighton commented:
@DdenBraver @HemantMahawar  @TravisEz13 Agreed; putting the original xDNSARecord resource back untouched with a warning will be the safest and easiest solution.

The composite resource suggestion would ensure xDNSARecord resource functionality would continue to work without breaking anything _and_ inherit any future improvements in the xDNSRecord resource without any further changes, tests or modifications! However, it's still a change.

DdenBraver commented:
@iainbrighton @HemantMahawar Thank you for all the feedback, and comments on the code.
I will do my best to edit and test the code this weekend.

DdenBraver commented:
@TravisEz13 Thank you, I somehow always struggle with either UTF8 or Unicode errors.

@HemantMahawarI updated the resource using your feedback, however do you have any hint where I can place the warning message so that it will show on module load? Currently I just added it to the get-targetresource so that it will show every time when used.

@iainbrighton I added the "Ensure" parameter to the DNSRecord resource now as well.

DdenBraver commented:
@TravisEz13 In this case, I placed it into the Get-TargetResource since in this case it is called by the Test-TargetResource.
If you still think it would be best to move it, please let me know.

iainbrighton commented:
Cool! Does it support creating multiple A records for round-robin operation? I think I logged that as a separate issue?

DdenBraver commented:
@iainbrighton Just did a few tests, and fixed the multiple A records support.

iainbrighton commented:
You star :)
